### DISCUSSION TITLE
SQL Query Dialect Conversion/Translation

### DISCUSSION THREAD

**Discussion by jlynchMicron** [sqlalchemy/sqlalchemy#9162] [Ideas]
Hi All,

This may be a long shot, but what do people think of the feasibility/utility of having a SQL query dialect converter?  The use case I am foreseeing with this is mostly transferring queries between "production" remote databases and local SQLite databases.  I am in the position where I have to make an offline version of my application and I would like to transfer my SQL query logic to this offline application (via an SQLite subset version of the table) without having to rewrite all my queries.

I know this is probably a pretty niche use case, but I figure with the amount of query syntax parsing SQLalchemy does, it would be the best project to have that functionality added.

The only software I have found that purports to do this is JOOQ (https://www.jooq.org/translate/) but it appears to work through internet API calls which my application will not be able to do.

**Comment by zzzeek** [comment]
applications that talk to both PostgreSQL and SQLite (or SQL Server, MySQL, Oracle) without changing any code are commonplace.  that's one of several key rationales one would be using the [Core SQL Expression language](https://docs.sqlalchemy.org/en/20/core/sqlelement.html) for.   the "translation" is done by allowing the Python DSL-constructed statements compile on the fly to the target backend (or backends) in use.

**Reply by jlynchMicron** [reply]
Ah I guess that makes sense.  What if your queries are already written in plain text for one of the dialects?  

Is there a way to:

1. Compile the text with a dialect. Example: sa.text(QUERY_TEXT).compile(dialect=postgresql)
2. Reverse the compilation to a general (ClauseElement?) representation.
3. Re-compile with a different dialect.

**Reply by CaselIT** [reply]
No, sqlalchemy has no sql parser, so it has no way of changing a sql string to adapt it from one dialect to another

**Reply by jlynchMicron** [reply]
Gotcha, well thanks for the info!  I will probably just adapt my old text queries to SQLalchemy expressions.
